Add AudioFlamingo3 model support

lashahub opened the pull request:
<!-- markdownlint-disable -->
## Purpose

This PR adds support for the **AudioFlamingo3** model family (e.g., [`nvidia/audio-flamingo-3-hf`](https://huggingface.co/nvidia/audio-flamingo-3-hf)).

[AudioFlamingo3](https://huggingface.co/docs/transformers/en/model_doc/audioflamingo3) is a large audio-language model. This implementation includes:
- The model executor definition.
- A custom multimodal processor (`AudioFlamingo3MultiModalProcessor`) to handle audio inputs and specific chunking logic.
- Support for `llm.chat()` with audio inputs via `audio_url`.
- Unit and integration tests.

## Test Plan

I have added two new test files to verify the implementation:

1. **Unit Tests**: Verify the processor's chunking logic and dummy data generation.
   ```bash
   pytest tests/models/multimodal/processing/test_audioflamingo3.py
   ```

2. **Generation Tests**: Verify end-to-end generation against official fixtures.
   ```bash
   pytest tests/models/multimodal/generation/test_audioflamingo3.py
   ```

## Test Result

- **Processing Tests**: Passed. Verified that audio chunk counts are calculated correctly based on duration.
- **Generation Tests**: Passed. The outputs match the expected results defined in the Hugging Face Transformers fixtures (using `expected_results_single.json` and `expected_results_batched.json`), ensuring parity with the HF implementation.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

lashahub committed:
Add Audio Flamingo 3 model support

Signed-off-by: Lasha <26011196+lashahub@users.noreply.github.com>

lashahub committed:
Add tests & fixtures

Signed-off-by: Lasha <26011196+lashahub@users.noreply.github.com>

jeejeelee commented:
To add a new Audio LLM model, you also need to :
- Add the model to the [documentation](https://github.com/vllm-project/vllm/blob/main/docs/models/supported_models.md#generative-models-1)
- Register the model in [test registry](https://github.com/vllm-project/vllm/blob/main/tests/models/registry.py#L578)
- Add the model to the [audio examples](https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/audio_language.py)

lashahub committed:
Enable LoRA support, add examples, and update docs

Signed-off-by: Lasha <26011196+lashahub@users.noreply.github.com>

DarkLight1337 commented on examples/offline_inference/audio_language.py:
```diff
@@ -99,6 +99,34 @@ def run_voxtral(question: str, audio_count: int) -> ModelRequestData:
     )
 
 
+# AudioFlamingo3
```
Feel free to move voxtral

jeejeelee commented on vllm/model_executor/models/audioflamingo3.py:
```diff
@@ -0,0 +1,637 @@
+    dummy_inputs=AudioFlamingo3DummyInputsBuilder,
+)
+class AudioFlamingo3ForConditionalGeneration(
+    nn.Module, SupportsMultiModal, SupportsPP, SupportsLoRA
```
Regarding MM models, to support LoRA, the model must implement `get_mm_mapping`. You can implement this function by referring to [ultravox](https://github.com/vllm-project/vllm/blob/v0.12.0/vllm/model_executor/models/ultravox.py).

lashahub committed:
Set min Transformers version to 5.0.0.dev for AF3

Co-authored-by: Isotr0py <2037008807@qq.com>
Signed-off-by: Lasha Koroshinadze <26011196+lashahub@users.noreply.github.com>

lashahub committed:
Alphabetical order in registry & ruff format

Signed-off-by: Lasha <26011196+lashahub@users.noreply.github.com>

lashahub committed:
Add get_mm_mapping to AF3

Signed-off-by: Lasha <26011196+lashahub@users.noreply.github.com>

lashahub committed:
Add multi-audio support flag

Co-authored-by: Isotr0py <2037008807@qq.com>
Signed-off-by: Lasha Koroshinadze <26011196+lashahub@users.noreply.github.com>

lashahub commented:
Thanks for the reviews! @jeejeelee @Isotr0py @DarkLight1337 

Changes:
1. Added `get_mm_mapping` to correctly target the language model for LoRA adapters.
2. Alphabetized registry entries.
3. Updated supported models table to reflect multi-audio capabilities (`T + A+`).

DarkLight1337 reviewed (approved):
LGTM now

jeejeelee committed:
Merge branch 'main' into main

lashahub committed:
Optimize AudioFlamingo3Encoder forward pass

Signed-off-by: Lasha <26011196+lashahub@users.noreply.github.com>

DarkLight1337 committed:
Merge branch 'main' into main

lashahub commented:
Update: batch `input_features` in `AudioFlamingo3Encoder` to improve performance and remove code duplication, as suggested by _Gemini Code Assist_

DarkLight1337 commented on tests/models/multimodal/generation/test_audioflamingo3.py:
```diff
@@ -0,0 +1,137 @@
+# SPDX-License-Identifier: Apache-2.0
```
Can you skip this file or check against the transformers version in `HF_EXAMPLES` then?

lashahub committed:
Skip AF3 tests if transformers<5.0.0.dev0

Signed-off-by: Lasha <26011196+lashahub@users.noreply.github.com>
